cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/DefaultKafkaQuotasManager.java

```diff
@@ -0,0 +1,210 @@
+        }
+    }
+
+    /**
+     * Checks whether default quotas configuration needs to be altered with new configuration.
+     * It gets the current default quotas for users from Kafka and compares them with the desired configuration
+     *
+     * @param reconciliation          Reconciliation marker
+     * @param vertx                   Vert.x instance
+     * @param kafkaAdmin              Kafka Admin object
+     * @param ops                     List of {@link ClientQuotaAlteration.Op}
+     * @param isNotKafkaPlugin        boolean parameter determining if the Kafka built-in quotas plugin is used
+     *
+     * @return  result determining if the new quotas configuration should be applied or not
+     */
+    private static Future<Boolean> shouldAlterDefaultQuotasConfig(Reconciliation reconciliation, Vertx vertx, Admin kafkaAdmin, List<ClientQuotaAlteration.Op> ops, boolean isNotKafkaPlugin) {
+        return VertxUtil.kafkaFutureToVertxFuture(reconciliation,
+                vertx,
+                kafkaAdmin.describeClientQuotas(ClientQuotaFilter.containsOnly(List.of(ClientQuotaFilterComponent.ofDefaultEntity(ClientQuotaEntity.USER)))).entities())
+            .compose(clientQuotaEntityMapMap -> {
+                if (clientQuotaEntityMapMap != null) {
+                    Map<String, Double> currentQuotas = clientQuotaEntityMapMap.get(DEFAULT_USER_ENTITY);
+                    Map<String, Double> desiredQuotas = new HashMap<>();
+                    ops.forEach(op -> {
+                        // the result of `describeClientQuotas` contains just non-null values, in case that we will have
+                        // null values in desiredQuotas, the Maps will never be equal and the quotas will be altered every
+                        // reconciliation (even if the desired and current values will be same)
+                        if (op.value() != null) {
+                            desiredQuotas.put(op.key(), op.value());
+                        }
+                    });
+
+                    // 1. if the current quotas set in Kafka are null and quotas configuration in Kafka CR is not of type `QuotasPluginKafka`, skip alteration
+                    // 2. if the current quotas set in Kafka are null and quotas configuration in Kafka CR is type of `QuotasPluginKafka`, but there is no fields configured, skip alteration
+                    if ((currentQuotas == null && isNotKafkaPlugin) || (currentQuotas == null && desiredQuotas.isEmpty())) {
+                        LOGGER.debugCr(reconciliation, "There are no default user quotas set in Kafka and the Kafka built-in plugin is not configured, skipping the alteration");
+                        return Future.succeededFuture(false);
+                    }
+
+                    return Future.succeededFuture(!desiredQuotas.equals(currentQuotas));
+                } else {
+                    return Future.failedFuture("Failed to get current default user quotas from Kafka");
```

Comment:
What are the situations when you get this?

Reply:
I think that some KafkaException would be thrown in case that the describe operation for the quotas will fail. I wanted to have there some null check and prevent NPE in case that it will be really `null`, so handling it like that looked good to me.
However I can remove it, because I think that it will either be filled with single record (that we are finding), where the value will be null (already checked in the first if), or an Exception will be thrown.

Reply:
I guess you should have a clear ide when it might be a null and whether that really indicates that you failed to get something. Returning successful future with null normally doesn't indicate any problems but for example that the record you looked for does not exist. If that is the case, you should know quite exactly when it happens and why. So you should cover that in the error message. If the null is returned really from Kafka to indicate some error, then fine, but you should explain it in a comment to make it clear.

Reply:
Yeah as I mentioned, the result of the `describeClientQuotas` is Map containing another Map based on the filter you apply there - even if the value for such filter doesn't exist, it will contain the key (of the entity we are looking for) and the value -> which can be either `null` (no quota exists for such entity) or Map with keys (names of quotas) and their values.

Only when there is issue with describing the quotas (for example the broker is down), we get the KafkaException -> but that is thrown on different place and the null check there is not catching it.

So I guess it's not needed to have the failed future there.